### PR TITLE
Combine reviewer summaries and fix all open findings

### PR DESCRIPTION
**Pull request by anish-inf**
## Summary

- keep configured reviewers running in parallel, but publish one combined PR summary
- make bare `@loupe fix` collect every unresolved finding from configured reviewers and address them in one agent run and commit
- acknowledge fix commands immediately, update progress with finding/reviewer counts, and report completion
- guard fix pushes against concurrent PR-head changes
- preserve reviewer-specific markers for incremental review state

## Validation

- `bunx oxlint --type-aware packages`
- `bun run tsc`
- `bun test`
- exercised end to end in `anish-inf/loupe-sandbox#2`: 13 findings from 3 reviewers were collected and fixed in commit `080cff2`

## Sandbox

https://github.com/anish-inf/loupe-sandbox/pull/2

---

> [!NOTE]
> **Medium Risk**
> Changes GitHub comment layout and chat fix/push behavior; mistakes could affect incremental review markers or push safety on active PR branches.
> 
> **Overview**
> **Loupe now publishes one combined PR summary** instead of separate per-reviewer issue comments. Reviewers still run in parallel and post their own inline reviews, but the action sets `deferSummary`, aggregates each reviewer’s `summaryBody` into a single comment (`<!-- loupe:summary:combined -->`), and preserves per-reviewer SHA markers for incremental state. Skipped reviewers keep their prior section; legacy per-reviewer summaries are tagged stale after migration.
> 
> **`@loupe fix` without arguments** (or `@loupe fix all`) gathers unresolved Loupe inline threads for configured reviewers via GraphQL (`listOpenLoupeFindings`), runs one fixer agent with `buildFixFindingsUserPrompt`, and commits once. Targeted `@loupe fix <what>` is unchanged. Fix flow adds ack/progress comments, refuses when there are no open findings, and **re-checks PR head** before checkout and again before push so it won’t overwrite newer commits.
> 
> Docs and README describe the combined summary and the new fix behavior. Core/action tests cover open-finding collection and combined-summary edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit c8da935f7a5286ea6a410266f2078ff515c3b7d1. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/action/src/orchestrate.ts [hunk 1]
```diff
@@ -55,6 +60,28 @@ async function runOne(
  * review regardless of config. Reviewer failures are reported on the PR here
  * and returned as outcomes. Only setup errors (bad config) reject.
  */
+function renderCombinedSummary(outcomes: readonly ReviewerOutcome[]): string {
+  const sections = outcomes.map((outcome) => {
+    if (!outcome.ok) {
+      return `## ${outcome.name}\n\n⚠️ Reviewer failed: ${outcome.error.split("\n")[0]?.slice(0, 300)}`;
+    }
+    if (!outcome.result.summaryBody) {
+      return `## ${outcome.name}\n\n_Not run: ${outcome.result.summary}_`;
```
[the remark this answers is not in the document]
Comment (anish-inf) [reply packages/action/src/orchestrate.ts]: Fixed in 3c75087. Skipped incremental reviewers now retain their prior full summary section and SHA marker, with a visible “Not updated in this run” note. Added a regression test for preservation.

File: packages/core/src/github.ts [hunk 2]
```diff
@@ -303,17 +326,75 @@ query LoupeReviewThreads($owner: String!, $repo: String!, $number: Int!, $after:
...
+    for (const thread of conn.nodes) {
+      if (!thread || thread.isResolved) continue;
+      const root = thread.comments.nodes[0];
+      const rootAuthor = root?.author?.login;
+      const authoredBySelf =
+        rootAuthor === self ||
+        (self === "github-actions[bot]" && rootAuthor === "github-actions");
+      if (!root || root.replyTo || !authoredBySelf) continue;
+      const marker = parseFindingMarker(root.body);
+      if (!marker || marker.sha !== headSha) continue;
```
[the remark this answers is not in the document]
Comment (anish-inf) [reply packages/core/src/github.ts]: Fixed in 3c75087. Bare `@loupe fix` now includes unresolved Loupe threads from earlier incremental heads for configured reviewers, while still excluding resolved, human-authored, and unconfigured-reviewer threads. Added regression coverage.

File: packages/core/src/github.ts [hunk 3]
```diff
@@ -654,30 +828,33 @@ export async function postReview(
...
+  if (!opts.deferSummary) {
+    const self = await getSelfLogin(octokit);
+    const comments = await listIssueComments(octokit, ref);
+    const priorSummary = comments
+      .reverse()
+      .find(
+        (comment) =>
+          comment.user?.login === self &&
+          comment.body?.includes(summaryMarkerPrefix(opts.reviewerName)),
+      );
```
[the remark this answers is not in the document]
Comment (anish-inf) [reply packages/core/src/github.ts:906]: Fixed in 63f3847. Legacy per-reviewer lookup now explicitly excludes the combined summary and stale-tagged comments.

File: packages/core/src/github.ts [hunk 4]
```diff
@@ -595,8 +677,100 @@ export type PostReviewOptions = {
...
+      comment.body?.includes("<!-- loupe:summary:") &&
+      !comment.body.includes(COMBINED_SUMMARY_MARKER)
+    ) {
+      await octokit.issues.deleteComment({
+        owner: ref.owner,
+        repo: ref.repo,
+        comment_id: comment.id,
+      });
+    }
+  }
```
[the remark this answers is not in the document]
Comment (anish-inf) [reply packages/core/src/github.ts:835]: Fixed in 63f3847 with the non-destructive migration behavior we chose: legacy summaries are preserved and tagged `loupe:summary:stale`; each update is best-effort and cannot fail the review.

File: packages/core/src/github.ts [hunk 5]
```diff
@@ -595,8 +677,100 @@ export type PostReviewOptions = {
...
+  priorBody?: string,
+): string {
+  if (!priorBody) return body;
+  return body.replace(
+    /## ([^\n]+)\n\n_Not run: [^\n]*_(?=\n\n---|$)/g,
+    (stub, reviewer: string) => {
+      const escaped = reviewer.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
+      const priorSection = new RegExp(
+        `## ${escaped}\\n\\n[\\s\\S]*?(?=\\n\\n---|\\n\\n<!-- loupe:summary:combined -->|$)`,
+      ).exec(priorBody)?.[0];
```
[the remark this answers is not in the document]
Comment (anish-inf) [reply packages/core/src/github.ts]: Fixed in 63f3847. Restoration is now bounded by the trusted reviewer SHA marker instead of Markdown horizontal rules; regression coverage includes an internal `---`.

File: packages/action/src/orchestrate.ts [hunk 6]
```diff
@@ -182,5 +209,10 @@ export async function runReviews(
     });
   }
 
+  await upsertCombinedSummary(
+    makeOctokit(config.token, logger),
+    { owner: config.owner, repo: config.repo, pull_number: config.pullNumber },
+    renderCombinedSummary(outcomes),
+  );
```
[the remark this answers is not in the document]
Comment (anish-inf) [reply packages/action/src/orchestrate.ts]: Fixed in 63f3847. The original `@loupe review` acknowledgement is now updated to a terminal warning state when combined-summary publication throws, and the job remains marked failed.

File: packages/core/src/github.ts [hunk 7]
```diff
@@ -595,8 +677,119 @@ export type PostReviewOptions = {
...
+  const escaped = reviewer.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
+  const marker = new RegExp(
+    `<!-- loupe:summary:${escaped} sha=[0-9a-f]{7,40} -->`,
+  ).exec(priorBody);
+  if (!marker?.[0] || marker.index === undefined) return undefined;
+  const heading = `## ${reviewer}\n\n`;
+  const start = priorBody.lastIndexOf(heading, marker.index);
+  if (start < 0) return undefined;
+  return priorBody.slice(start, marker.index + marker[0].length);
+}
```
[the remark this answers is not in the document]
Comment (anish-inf) [reply packages/core/src/github.ts:724]: Fixed in the latest push. Combined summaries now wrap each reviewer in explicit hidden start/end markers. Legacy restoration is bounded by the next generated heading/footer and only accepts a SHA marker inside those bounds, so a retained marker cannot absorb later reviewers. Added a regression test for this exact shape.

File: packages/action/src/respond.ts [hunk 8]
```diff
@@ -242,25 +272,85 @@ export async function handleComment(
...
-      await postFailure(octokit, ref, "review", err, logger);
+      const reason = err instanceof Error ? err.message : String(err);
+      logger.error("Chat command failed: review", { error: reason });
+      process.exitCode = 1;
+      await updateIssueComment(
+        octokit,
+        ref,
+        ackId,
+        `⚠️ Re-review finished, but Loupe could not publish the combined summary — ${reason.slice(0, 500)}\n\nSee the Actions run logs for details.`,
+      );
```
[the remark this answers is not in the document]
Comment (anish-inf) [reply packages/action/src/respond.ts:289]: Fixed in the latest push. `runReviews` now wraps only combined-summary publication failures in `CombinedSummaryPublicationError`; the chat ack uses the specific summary message only for that type and a generic incomplete-review message for setup/other failures.

File: packages/core/src/github.ts [hunk 9]
```diff
@@ -595,8 +677,146 @@ export type PostReviewOptions = {
...
+  const candidates = [nextHeading, footer, combinedMarker].filter(
+    (index) => index >= 0,
+  );
+  const end =
+    candidates.length > 0 ? Math.min(...candidates) : priorBody.length;
+  const section = priorBody.slice(start, end).trim();
+  const marker = new RegExp(
+    `<!-- loupe:summary:${escaped} sha=[0-9a-f]{7,40} -->`,
+  ).exec(section);
+  return marker ? section : undefined;
```
[the remark this answers is not in the document]
Comment (anish-inf) [reply packages/core/src/github.ts:723]: Fixed in c8da935. Marked combined summaries are now processed structurally: Loupe detects a skipped state inside each reviewer’s explicit start/end pair and replaces only that pair’s content with the prior bounded section. The new boundaries remain intact. Legacy unmarked restoration remains as a separate fallback. Added tests for marked restoration, one boundary pair, no later-reviewer absorption, and legacy behavior.
